src/boot/efi/boot.c

```diff
@@ -1624,12 +1633,36 @@ static INTN config_entry_compare(const ConfigEntry *a, const ConfigEntry *b) {
         assert(a);
         assert(b);
 
-        /* Order entries that have no tries left to the beginning of the list */
-        if (a->tries_left != 0 && b->tries_left == 0)
-                return 1;
+        /* Order entries that have no tries left to the end of the list */
         if (a->tries_left == 0 && b->tries_left != 0)
+                return 1;
+        if (a->tries_left != 0 && b->tries_left == 0)
                 return -1;
 
+        /* First, order by sort key (alphabetically upwards), if it is defined. */
+        if (a->sort_key && b->sort_key) {
+                r = strcmp(a->sort_key, b->sort_key);
+                if (r != 0)
+                        return r;
+
+                /* If multiple installations of the same OS are around, group by machine ID */
+                if (a->machine_id && b->machine_id) {
+                        r = strcmp(a->machine_id, b->machine_id);
```

Comment:
here too

Reply:
as above: we should only compare by machine ID if both entries have one set. If one doesn't this is a reason to try something else, but not a reason to order the one without before the one with